deepspeed/utils/zero_to_fp32.py

```diff
@@ -398,9 +397,71 @@ def _zero3_merge_frozen_params(state_dict, world_size, zero_model_states):
     print(f"Reconstructed Frozen fp32 state dict with {total_params} params {total_numel} elements")
 
 
+class GatheredTensor:
+    """
+    A pseudo tensor that collects partitioned weights.
+    It is more memory efficient when there are multiple groups.
+    """
+    def __init__(self, flat_groups, offset, partitioned_numel, shape):
+        self.flat_groups = flat_groups
+        self.offset = offset
+        self.partitioned_numel = partitioned_numel
+        self.shape = shape
+
+    def contiguous(self):
+        """
+        Merge partitioned weights from flat_groups into a single tensor.
+        """
+        end_idx = self.offset + self.partitioned_numel
+        world_size = len(self.flat_groups)
+        pad_flat_param_chunks = []
+        for rank in range(world_size):
+            flat_group = self.flat_groups[rank]
+            tensor_slice = []
+            start_part = None
+            end_part = None
+            sum_idx = [0] + list(np.cumsum([flat_tensor.numel() for flat_tensor in flat_group]))
+            for i in range(len(sum_idx)):
+                if sum_idx[i] <= self.offset < sum_idx[i + 1]:
+                    start_part = i
+                if sum_idx[i] < end_idx <= sum_idx[i + 1]:
+                    end_part = i
+                    break
+            for i in range(start_part, end_part + 1):
+                flat_tensor = flat_group[i]
+                start_offset = self.offset - sum_idx[i]
+                end_offset = min(end_idx, sum_idx[i + 1]) - sum_idx[i]
+                tensor_slice.append(flat_tensor[start_offset:end_offset])
+
+            pad_flat_param_chunks.append(torch.concat(tensor_slice, 0))
+        pad_flat_param = torch.cat(pad_flat_param_chunks, dim=0).to(torch.float16)
```

Comment:
Why is this cast to fp16? The checkpoint could be bf16.

Reply:
to fp16 has been removed